### DISCUSSION TITLE
Dynamic routing

### DISCUSSION THREAD

**Discussion by abhishekDev047** [vercel/next.js#47947] [Help]
### Summary

hey everyone I'm a beginner web developer, I've just switched to next.js from reacct.js and I have a question. why we are confusingly doing dynamic routing, we are exporting functions from here to there and then using them in the other component, and at last, we are just going to the, for example, posts/ssr.md file. if we are directly jumping to that exact path then why do we need to follow that confusing process? I have given the link below in the example tab

### Example

https://nextjs.org/learn/basics/dynamic-routes/implement-getstaticpaths

**Comment by icyJoseph** [comment]
Hi,

Next.js is a React framework, so no change there. You are still using React, within the Next.js production framework.

The way dynamic routing is done in your example is making use a of a Next.js feature called: data fetching functions.

At the top of a page component, we can define a few function combinations:

- getStaticProps + getStaticPaths
- getStaticProps
- getServerSideProps

If I understand correctly, your confusion is on why do we create a function to: `getAllPostsIds`, and then a function to `getPostData`, both of which just read file system, and then import them into the page component (dynamic route), when we could just inline everything?

Well, it won't always be the case that these are available in the file system. In most cases this data lives in a CMS, so you have to learn to structure your data-fetching so that the details of implementation do not pollute the data injection process.

In other words, you could change `getAllPostsIds` internal implementation for a call to a CMS, and nothing in the dynamic route code itself would have to change.

This also helps the tutorial have more clarity in the `getStaticPaths` and `getStaticProps` intend, which is to "somehow" get data and return it, so that it can be used by the framework.
